lib/spack/spack/environment.py

```diff
@@ -0,0 +1,200 @@
+import os
+import os.path
+import collections
+
+
+class AttributeHolder(object):
+    """
+    Policy that permits to store any kind of attribute on self. The attributes must be passed as key/value pairs
+    during the initialization of the instance.
+    """
+    def __init__(self, **kwargs):
```

Comment:
Are the `kwargs` here ever used?  I don't think so -- maybe this is too generic?

I like having a class per type-of-modification, but I think you want something more like this:

```python
class NameModifier(object):
    def __init__(self, name):
        self.name = name

class NameValueModifier(object):
    def __init__(self, name, value):
        self.name = name
        self.value = value

class SetEnv(NameValueModifier):
    def execute(self):
        os.environ[self.name] = str(self.value)

class UnsetEnv(NameModifier):
    def execute(self):
        os.environ.pop(self.name, None)
```

Now the subclasses are much shorter and more concise...